cranelift/codegen/src/isa/x64/inst/args.rs

```diff
@@ -0,0 +1,434 @@
+            ExtMode::WL => 4,
+            ExtMode::WQ => 8,
+            ExtMode::LQ => 8,
+        }
+    }
+}
+
+impl fmt::Debug for ExtMode {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.to_string())
+    }
+}
+
+// These indicate the form of a scalar shift: left, signed right, unsigned
+// right.
+#[derive(Clone)]
+pub enum ShiftKind {
+    Left,
+    RightZ,
+    RightS,
+}
+
+impl ShiftKind {
+    pub(crate) fn to_string(&self) -> String {
+        match self {
+            ShiftKind::Left => "shl".to_string(),
+            ShiftKind::RightZ => "shr".to_string(),
+            ShiftKind::RightS => "sar".to_string(),
+        }
+    }
+}
+
+impl fmt::Debug for ShiftKind {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.to_string())
+    }
+}
+
+// These indicate condition code tests.  Not all are represented since not all
+// are useful in compiler-generated code.
+#[derive(Copy, Clone)]
+pub enum CC {
+    O,   //  0   overflow
+    NO,  //  1   no overflow
+    B,   //  2   <u
+    NB,  //  3   >=u
+    Z,   //  4   zero
+    NZ,  //  5   not zero
+    BE,  //  6   <=u
+    NBE, //  7   >u
+    S,   //  8   negative
+    NS,  //  9   not negative
+    L,   //  12  <s
+    NL,  //  13  >=s
+    LE,  //  14  <=s
+    NLE, //  15  >s
+}
+
+impl CC {
+    pub(crate) fn to_string(&self) -> String {
+        match self {
+            CC::O => "o".to_string(),
+            CC::NO => "no".to_string(),
+            CC::B => "b".to_string(),
+            CC::NB => "nb".to_string(),
+            CC::Z => "z".to_string(),
+            CC::NZ => "nz".to_string(),
+            CC::BE => "be".to_string(),
+            CC::NBE => "nbe".to_string(),
+            CC::S => "s".to_string(),
+            CC::NS => "ns".to_string(),
+            CC::L => "l".to_string(),
+            CC::NL => "nl".to_string(),
+            CC::LE => "le".to_string(),
+            CC::NLE => "nle".to_string(),
+        }
+    }
+
+    pub(crate) fn invert(&self) -> CC {
+        match self {
+            CC::O => CC::NO,
+            CC::NO => CC::O,
+            CC::B => CC::NB,
+            CC::NB => CC::B,
+            CC::Z => CC::NZ,
+            CC::NZ => CC::Z,
+            CC::BE => CC::NBE,
+            CC::NBE => CC::BE,
+            CC::S => CC::NS,
+            CC::NS => CC::S,
+            CC::L => CC::NL,
+            CC::NL => CC::L,
+            CC::LE => CC::NLE,
+            CC::NLE => CC::LE,
+        }
+    }
+
+    pub(crate) fn get_enc(&self) -> u8 {
+        match self {
+            CC::O => 0,
+            CC::NO => 1,
+            CC::B => 2,
+            CC::NB => 3,
+            CC::Z => 4,
+            CC::NZ => 5,
+            CC::BE => 6,
+            CC::NBE => 7,
+            CC::S => 8,
+            CC::NS => 9,
+            CC::L => 12,
+            CC::NL => 13,
+            CC::LE => 14,
+            CC::NLE => 15,
+        }
```

Comment:
```suggestion
        self as u8
```

Reply:
I'd prefer to retain the original. That makes it easy to figure out the encoding for each condition. With `self as u8` you have to count through the definition to find the encoding.

Reply:
The previous suggestion explicitly assigns a discriminant to every variant. Otherwise this wouldn't even be correct, as 10 and 11 are skipped.